### DISCUSSION TITLE
Add support for nightlight on the Chime Pro

### DISCUSSION THREAD

**Discussion by jampez77** [tsightler/ring-mqtt#528] [Feature Requests]
I recently got a [Chime Pro ](https://www.amazon.co.uk/dp/B087Q2MD5N?psc=1&ref=ppx_yo2ov_dt_b_product_details) that has a night light.

The ring app gives the option to turn it on or off but if possible I'd prefer to do so using this project so I can automate the light with the rest of my smart home.

**Comment by tsightler** [comment]
Biggest problem is, as it's not directly supported by ring-client-api, and I don't know the command that is needed for this (the URL or any json body specs) so I'm not sure how to implement it.  Normally you can somewhat reverse engineer the commands from the web UI, but chimes aren't available there, as far as I know (at least my Chime Pro doesn't appear there).  That means you'd have to deconstruct the Ring app or trace traffic with a MITM proxy setup (and probably a rooted device) and I don't have that setup these days, nor the time to do so.  Do you know if these commands are documented in any other Ring API project?

**Reply by tsightler** [reply]
Took a look at python-ring-doorbell project at it seems there is also a [feature request](https://github.com/tchellomello/python-ring-doorbell/issues/263) there, but no response so I'm guessing no one has yet done the work to reverse engineer the required call.    I do happen to have a Chime Pro so I can take a blind stab at it or two, but that has low chance of success I'd guess, unless Ring reused the normal light on/off, which maybe we'll get lucky and they did.

**Reply by jampez77** [reply]
A stab in the dark is always worth a try. I've downloaded 'mitmproxy' and will try and give something a go after the weekend if I can.

Hopefully someone figures something out on that other feature request.

Thanks for looking into this!

**Reply by tsightler** [reply]
I took quite a few stabs but didn't find the magic combination.  Until someone manages to reverse engineer the required commands, I don't think I can do anything here.  I'd try to do that myself, but I just won't have time for the forseeable future.

**Reply by jampez77** [reply]
Ahh fair enough. Like I say I will give some reverse engineering all with mitmproxy next week but it's the first time I've looked at anything of the sort so I wouldn't hold your breathe haha

**Reply by jampez77** [reply]
I gave this a go today on my lunch and it's fair to say I don't know what i'm doing as I couldn't get the app to work with the proxy setup (on an iphone).

**Reply by tsightler** [reply]
I doubt you'd be able to do it from an iPhone.  I believe the Ring app uses pinned certificates so probably need a root Andriod (perhaps emulated) so that you can remove the pinned cert and add the custom cert for MITM proxy.  I haven't done it since like 2019, so not sure what all is required these days, but MITM website is pretty well documented, it's just a lot of steps.

**Reply by jampez77** [reply]
Ok so I think I might have something! I have managed to get MITM working via an emulator! Thanks for the suggestion. 

I don't know exactly what you need so i will just dump the screen shots here...

When I replay the requests I can see the nightlight being updated in the app. Obviously `light_sensor_enabled` needs to be set to true or false.

<img width="638" alt="Screenshot 2023-01-18 at 09 52 36" src="https://user-images.githubusercontent.com/1823647/213140486-87b2a3e3-9549-4b3b-9bb3-14568f01e170.png">
<img width="639" alt="Screenshot 2023-01-18 at 09 53 01" src="https://user-images.githubusercontent.com/1823647/213140492-0855e859-24b1-4fbe-bec9-a94a471add01.png">

**Reply by jampez77** [reply]
FYI This is how I got the emulator setup

First I created an emulator (Android 12) and set up the proxy info as described here:

https://medium.com/@navyab_66433/mitm-proxy-for-android-emulators-cf4c8e909aac

You might also need to use the accepted answer on this SO question so you can configure the proxy
https://stackoverflow.com/questions/70972106/how-to-configure-proxy-in-emulators-in-new-versions-of-android-studio

At this stage I still couldn't see any traffic due to the certificate pinning. But then I found this amazing library

https://github.com/shroudedcode/apk-mitm

From what i understand it rewrites the network security config and other things and created a new APK to use. 

Viola it worked!

**Reply by jampez77** [reply]
On a closer look I can see the night_light_settings object has a luma_light_threshold. It would be interesting to have a play around and see if that has any effect on the night light coming on as the app mentions it is based on light levels.

**Reply by tsightler** [reply]
Thanks for the details.  I feel like that first example is pretty close to one of my stabs in the dark, but I never got it to work.  Maybe I had something that was close, but wasn't quite close enough.  It will probably be this weekend before I get to try it again, but that looks promising.

**Reply by jampez77** [reply]
I shall be crossing my fingers until then. 🤞

**Reply by tsightler** [reply]
I think I see why it my attempts didn't work.  I based my attempts in the updateChime() function in ring-client-api since that is what is used to set the volume and those prorperties also appear in settings, however, when I look at the data you provided, I see it uses the settings API URL which aligns more with the setDeviceSettings() function which ring-client-api uses for cameras.  This makes sense as camera and chimes share a lot of commonalities from an API perspective.  Unfortunately, ring-client-api doesn't implement this for chimes, but I can just call the API directly, so I think it should work.

**Reply by jampez77** [reply]
Ooo that's exciting. I'm increasingly hopeful. I think tomorrow I might have a look at the doorbell volume requests mentioned in another discussion.

**Reply by tsightler** [reply]
I can confirm that I've been able to enable/disable the nightlight via the API.  As I'm slowly nearing the release of the 5.1.0 version (currently testing in the dev branch) I will try to add this feature there over the weekend.  Thanks for your effort to flush out the API calls needed.

**Reply by jampez77** [reply]
That is excellent news! Did you happen to see if setting `luma_light_threshold` does anything?

**Reply by tsightler** [reply]
I didn't try it.  Is that setting available in the Ring app?  I don't remember seeing it.  In general, I'm pretty skeptical of implementing any code that attempts to change values via the API that aren't possible to change in the Ring app because there are LOTs of settings on the various devices that don't seem to work or be fully implemented, or at least don't work in expected ways (cameras have tons of such settings).

**Reply by jampez77** [reply]
Ah ok. No there is nothing in the app that I can see. I was just hoping that I stumbled upon a secret setting to change the nightlight illuminance threshold.

**Reply by tsightler** [reply]
Dev branch now has support for setting the nightlight enable status.  I exposed this as a switch to toggle between enabled/disabled, rather than a light since it's not possible to control the light directly, the toggle only enables/disables the light sensor.  There is also an NightlightState attribute available on the entity which can be used to determine the actual on/off state of the nightlight.

**Reply by jampez77** [reply]
That was quick! Good work. I will give the dev branch a go soon :)

**Reply by jampez77** [reply]
Forgot to say, this has been working flawlessly for me
